docs/dev/relay_pass_infra.rst

```diff
@@ -0,0 +1,616 @@
+    };
+    relay::transform::Pass seq = relay::transform::Sequential(pass_seqs);
+    
+    // Create a pass context for the optimization.
+    auto ctx = relay::transform::PassContext::Create();
+    ctx->opt_level = 2;
+    ctx->fallback_device = kDLCPU;
+
+    // Use the Python with syntax to execute the sequence of optimizations.
+    tvm::With<relay::transform::PassContext> scope(ctx);
+    mod = seq(mod);
+
+    // View the updated module.
+    LOG(INFO) << relay::AsText(mod) << std::endl;
+
+Other types of passes should be directly invoked for execution on a module. For
+example, users can directly apply const folding pass on a given module, ``mod
+= transform::FoldConstant()(mod)``. However, it is users' responsibility to
+execute the required passes explicitly.
+
+Pass Registration
+~~~~~~~~~~~~~~~~~
+
+We've covered the concept of different level of passes and the context used for
+compilation. It would be interesting to see how easily users can register
+a pass.  Let's take const folding as an example. This pass has already been
+implemented to fold constants in a Relay function (found in
+`src/relay/pass/fold_constant.cc`_).
+
+An API was provided to perform the ``Expr`` to ``Expr`` transformation.
+
+.. code:: c++
+
+    Expr FoldConstant(const Expr& expr);
+
+In order to register this pass to the pass infra, we first need to decide at
+which level this pass will be performed. As const folding happens on individual
+functions, we should intuitively create a ``FunctionPass`` for it through
+``CreateFunctionPass``. The ``pass_func`` is returned as a packed function that
+invokes the ``Expr`` to ``Expr`` API on each function in a Relay module. ``{}``
+indicates that no prerequisite is required for this pass. Otherwise, the pass
+developer has to identify and list them.
+
+Meanwhile, a pass API endpoint is registered with the name
+``relay._transform.FoldConstant``. This pass, therefore, becomes an entry in the
+registry that can be accessed by both C++ (e.g. the ``GetPass`` above) and
+Python when needed.
+
+.. code:: c++
+
+    namespace transform {
+
+    Pass FoldConstant() {
+      runtime::TypedPackedFunc<Function(Function, Module, PassContext)> pass_func =
+        [=](Function f, Module m, PassContext pc) {
+          return Downcast<Function>(FoldConstant(f));
+      };
+      return CreateFunctionPass(pass_func, 2, "FoldConstant", {});
+    }
+
+    TVM_REGISTER_API("relay._transform.FoldConstant")
+    .set_body_typed(FoldConstant);
+
+    }  // namespace transform
+
+To allow other C++ modules to apply this pass, we declare a free function in
+`include/tvm/relay/transform.h`_ as the following:
+
+.. code:: c++
+
+    TVM_DLL Pass FoldConstant();
+
+Python Frontend
+~~~~~~~~~~~~~~~
+
+Only some simple APIs are needed for the frontend side. For example, we can
+provide users the following APIs to create and execute a pass (full
+implementation is provided in `python/tvm/relay/transform.py`_). The backend
```

Comment:
I personally think "back-end" and "front-end" should be hyphenated but I don't know if everyone shares this opinion

Reply:
I think they're common enough words now that we can drop the hyphen.

Reply:
I will probably just keep them because it looks that they are also commonly used, including academic papers.